src/librustc_typeck/check/pat.rs

```diff
@@ -355,8 +355,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected: Ty<'tcx>,
         discrim_span: Option<Span>,
     ) -> Option<Ty<'tcx>> {
```

Comment:
I feel that if you change the return type to `Result<Ty<'tcx>, DiagnosticBuilder<'tcx>>` it will let you clean up a few lines below. Just make sure that in the single `check_pat_range` caller you `emit` the error.